src/cli/commands/run.ts

```diff
@@ -51,24 +52,42 @@ export function run() {
 
       logger.registerSynchronizationJobContext(synchronizationContext);
 
-      const executionResults = await executeIntegrationInstance(
-        logger,
-        createIntegrationInstanceForLocalExecution(invocationConfig),
-        invocationConfig,
-        {
-          enableSchemaValidation: true,
-        },
-      );
+      try {
+        const executionResults = await executeIntegrationInstance(
+          logger,
+          createIntegrationInstanceForLocalExecution(invocationConfig),
+          invocationConfig,
+          {
+            enableSchemaValidation: true,
+          },
+        );
 
-      log.displayExecutionResults(executionResults);
+        log.displayExecutionResults(executionResults);
 
-      await uploadCollectedData(synchronizationContext);
+        await uploadCollectedData(synchronizationContext);
 
-      const synchronizationResult = await finalizeSynchronization({
-        ...synchronizationContext,
-        partialDatasets: executionResults.metadata.partialDatasets,
-      });
+        const synchronizationResult = await finalizeSynchronization({
+          ...synchronizationContext,
+          partialDatasets: executionResults.metadata.partialDatasets,
+        });
+
+        log.displaySynchronizationResults(synchronizationResult);
+      } catch (err) {
+        await logger.flush();
+
+        if (!logger.isHandledError(err)) {
```

Comment:
I like this approach quite a bit actually, since we're using a stateful object as a logger 😄 It puts the responsibility in the right place I think, keeps us from having to keep track in the places where we do the logging by having to tag the error object.